File: src/skaters/runstats.py
"""Lightweight online running statistics (pure Python, no dependencies).

Welford's algorithm for mean/variance, plus exponentially weighted variants.
"""

from __future__ import annotations
import math


def running_var_init() -> dict:
    """Initialize state for Welford's online variance."""
    return {"n": 0, "mean": 0.0, "m2": 0.0}


def running_var_update(state: dict, x: float) -> dict:
    """Update running mean/variance with a new observation (Welford)."""
    n = state["n"] + 1
    delta = x - state["mean"]
    mean = state["mean"] + delta / n
    delta2 = x - mean
    m2 = state["m2"] + delta * delta2
    return {"n": n, "mean": mean, "m2": m2}


def running_var_get(state: dict) -> tuple[float, float]:
    """Return (mean, variance) from running stats state."""
    if state["n"] < 2:
        return state["mean"], float("inf")
    return state["mean"], state["m2"] / (state["n"] - 1)


def running_std_get(state: dict) -> float:
    """Return standard deviation from running stats state."""
    _, var = running_var_get(state)
    return math.sqrt(var) if math.isfinite(var) else float("inf")


def running_mse_get(state: dict) -> float:
    """Return mean squared error (second moment) from running stats of errors.

    MSE = bias² + variance, so it penalizes both systematic and random error.
    """
    if state["n"] < 1:
        return float("inf")
    mean, var = running_var_get(state)
    if not math.isfinite(var):
        return float("inf")
    return mean * mean + var

File: rust/src/runstats.rs
//! Welford online mean/variance. Port of src/skaters/runstats.py.

use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct RunningVar {
    pub n: u64,
    pub mean: f64,
    pub m2: f64,
}

impl RunningVar {
    pub fn new() -> RunningVar {
        RunningVar {
            n: 0,
            mean: 0.0,
            m2: 0.0,
        }
    }

    pub fn update(&mut self, x: f64) {
        let n = self.n + 1;
        let delta = x - self.mean;
        let mean = self.mean + delta / n as f64;
        let delta2 = x - mean;
        self.m2 += delta * delta2;
        self.n = n;
        self.mean = mean;
    }

    /// (mean, variance); variance is +inf until two observations.
    pub fn get(&self) -> (f64, f64) {
        if self.n < 2 {
            return (self.mean, f64::INFINITY);
        }
        (self.mean, self.m2 / (self.n - 1) as f64)
    }

    /// Mean squared error (bias^2 + variance) of tracked errors.
    pub fn mse(&self) -> f64 {
        if self.n < 1 {
            return f64::INFINITY;
        }
        let (mean, var) = self.get();
        if !var.is_finite() {
            return f64::INFINITY;
        }
        mean * mean + var
    }
}
